### PR TITLE
[NFC] Add utility function ur_unreachable to simplify code

### PR DESCRIPTION
**Pull request by wenju-he**

### CODE REVIEW

**Comment by wenju-he** [comment]
@kbenzie could you please take a look if this PR makes sense?

**Review by pbalcer — APPROVED** [review]

**Review by PietroGhg — APPROVED** [review]
Native CPU LGTM, thank you

File: source/common/ur_util.hpp [hunk 1]
```diff
@@ -295,4 +295,12 @@ inline ur_result_t exceptionToResult(std::exception_ptr eptr) {
 
 template <class> inline constexpr bool ur_always_false_t = false;
 
+[[noreturn]] inline void ur_unreachable() {
+#ifdef _MSC_VER
+    __assume(0);
+#else
+    __builtin_unreachable();
+#endif
+}
```
Comment (kbenzie) [inline source/common/ur_util.hpp:304]: I think we should put this in a namespace `ur::unreachable()`.

Comment (wenju-he) [reply source/common/ur_util.hpp]: done

**Review by kbenzie — APPROVED** [review]

**Comment by kbenzie** [comment]
I think there's a missing include in the CUDA file that's attempting to use `ur::unreachable`.

**Comment by wenju-he** [comment]
> I think there's a missing include in the CUDA file that's attempting to use `ur::unreachable`.

added to both cuda/usm.cpp and hip/usm.cpp

**Comment by kbenzie** [comment]
> > I think there's a missing include in the CUDA file that's attempting to use `ur::unreachable`.
> 
> added to both cuda/usm.cpp and hip/usm.cpp. I don't have access to cuda/hip machines to verify. After searching the headers used in hip/usm.cpp, it looks like ur_util.hpp is needed.

If you could add those includes we can let the CI check if the fix works.

**Comment by fabiomestre** [comment]
I have updated the target branch of this PR from the `adapters` branch to the `main` branch.
Development in UR is moving back to `main`. The `adapters` branch will soon be deleted.

**Review by JackAKirk — APPROVED** [review]

**Pull request merged**